tempesta_fw/http_limits.c

```diff
@@ -934,7 +934,12 @@ frang_http_req_process(FrangAcc *ra, TfwConn *conn, TfwFsmData *data,
 	T_FSM_INIT(Frang_Req_0, "frang");
 
 	BUG_ON(!ra);
-	BUG_ON(req != container_of(req->stream->msg, TfwHttpReq, msg));
+	/*
+	 * TODO: a NULL-dereference was triggered there for h2 messages,
```

Comment:
> if we receive RST_STREAM then the frame is processed and all the requests in it are unlinked and any next data frames should not be processed for the unlinked messages, i.e. we should not go into tfw_http_msg_process_generic()

Yes, that’s exactly how it should be: `req->stream` and `stream->msg` should be valid in any place of request receive path (until it passed to the server connection). This is controlled by stream FSM, and if there is a bug here, it is most believed that its causes are in the `tfw_h2_stream_fsm()` and the near code.
The check `(!req->stream)` in `tfw_h2_error_resp()` is made because this procedure can be called from response processing path.

Reply:
The real reason behind the issue was `tfw_http_sess_obtain()` call prior to frang call back. When sticky module builds a redirect message, it bounds a new response with the request, detach request from the stream: `tfw_h2_prep_redirect() -> tfw_h2_stream_id_close()`. 

I'm not sure, that we should call stream FSM at req<->resp binding stage, it's too early to call the egress fsm, which, I believe,  should be called right before pushing data into network.  But currently stream can be removed  at any time during request processing. That's the reason to use h2 context lock, when a stream is detached from the message.

But in my understanding, h2 stream is http message itself, so there is no need to destroy stream object at any time, it can be easier to destroy stream when request is destructed.

The code was changed a bit, now frang is called before sticky module. This much more helpful change that it seems: frang adds DDoS mitigation capabilities, and now it's called just when all the necessary information is available. Before this change attacker could send us tonns of messages without frang being involved (If all the messages wasn't split between multiple skbs). Http tables subsystem is still vulnerable for such attacks.